src/Selenium2Driver.php

```
@@ -352,7 +357,9 @@ public function start()
             $this->wdSession = $this->webDriver->session($this->browserName, $this->desiredCapabilities);
 
             $status = $this->webDriver->status();
-            $this->isW3C = version_compare($status['build']['version'], '3.0.0', '>=');
+            $majorSeleniumServerVersion = (int)explode('.', $status['build']['version'])[0];
+            $this->isW3C = $majorSeleniumServerVersion >= 3;
```

I see that we already have a case where we talk to selenium with W3C... we could solve the right-click problem in the same way: when w3c is enabled, we use W3C actions.

We could do that by somehow retrieving the http client the instaclick library is using (it seems they have some sort of DIC, so maybe not too complicated).

Give me a moment and I'll check if that's possible.

---

> - right click works in Selenium Server 2 and 4+

That is because selenium 4 correctly translates old/JWP right-click call to W3C actions.
If we make an action call instead, it will fix the problem for selenium 3 and selenium 4 will be functionally the same (at least from browser perspective).

> Because of this, we can't really say, that right-click doesn't work Selenium Server 3+ as you're suggesting.

I've a feeling you're replying to the wrong comment, or I misunderstood. 😆

---

> ... I see that we already have a case where we talk to selenium with W3C...

My apologies. The `$this->isW3C` code made you think, that we support W3C commands, but we don't.

As you can see we're just checking if the Selenium Server version supports W3C. This was needed to supply correct timeout names (`pageLoad` instead of `page load` based on the Selenium Server version).

As I've written somewhere else the `instaclick/web-driver` doesn't have the public methods for calling any non-implemented APIs as `php-webdriver/webdriver` does.

Since we agreed to not support right-clicking on Selenium Server 3 specifically (and this PR does just that) I recommend not spending any more time here. Especially since the `webdriver-classic-driver` is available with both right-clicking and Selenium Server 4 support.

If you still, for example, to practice, want to figure out how to use internal classes of the `instaclick/web-driver` to invoke the right-click action, then you're welcome.

---

It's not possible.

The same actions request by the classic driver will cause an UnknownCommandError response with this driver - I was stumped, considering that as said it works with the other driver.

The reason turned out to be simple:
- first of all, we need to keep in mind that the protocol (W3C vs JWP/OSS) applies in different places:
  1. it applies between PHP (our drivers) and Selenium
  2. it also applies between Selenium and the web browser
  3. additionally, since Selenium 3, selenium translates from JWP to W3C in case of point ii (and when the browser is also communicating with W3C).
- when we request a new session from selenium, it determines the type of protocol for both the driver side and for the web browser.
- in this particular case and with this driver's desiredCapabilities, selenium thinks the driver wants to communicate with JWP, therefore W3C API calls (between our driver and selenium) are disabled - causing the UnkownCommandError mentioned above
- in case of the classic driver, the JWP API calls are disabled and the W3C ones enabled (opposite situation)

---

This means that we do not have any other options - if the session is started in JWP, there's no way to switch to W3C as far as I can see.

---

> My apologies. The $this->isW3C code made you think, that we support W3C commands, but we don't.

Yeah, I was going to suggest maybe renaming that property, something like `useW3CStyleTimeout`, or since this whole thing is the situation for the right-click check, maybe just make one private property (for both cases) called `majorSeleniumServerVersion` or even just `isSelenium3`. It would make the intention clearer and less guesswork.

---

* The `$this->isW3C` checks that Selenum Server Version >= 3
* The `$this->isRightClickSupported` checks that Selenum Server Version != 3

I can't really merge them, because they check different things :(

Maybe they can be merged (not sure) IF this driver won't work with Selenium Server 4+. Will it work with Selenium Server 4+?

---

> Will it work with Selenium Server 4+?

Actually no, I happened to try testing with selenium 4, but the driver would not even connect.

So those two do kinda align because:
- `>=4` driver doesn't even connect (not even sure if status endpoint works*)
- `>=3` right click will not work
- `3` timeout requires(I guess?) the W3C format?

*if it does, I believe we should throw an early exception that Selenium4 is definitely not supported.

---

> `>=4` driver doesn't even connect (not even sure if status endpoint works*)

The `status` works. I've checked today.

It might even create a session if you wrap all the desired capabilities in the `firstMatch` element, but that's super non-obvious and would break browser-specific capabilities. I'll update the error message to say `Selenium Server 3` instead of `Selenium Server 3+`.

> *if it does, I believe we should throw an early exception saying that Selenium4 is definitely not supported.

Good catch.

The earliest things would be upon an attempt to create a new session. In fact, I can do a `status` API call even for Selenium 4 and it passes. I can use that to decide if an exception should be thrown.

---

I've made a commit with the proposed changes.

Please review.

P.S.
There are no tests currently to confirm, that:

1. session can't be created on Selenium Server > 3.x
2. doing a right click on Selenium Server 3.x would throw an exception

Should I add them?

Even if we add them, then how do I test on Selenium 4 so that only single test is executed?

---

> 1. session can't be created on Selenium Server > 3.x

Since you add a check that throws an exception for that, I would recommend adding a test - it should be easy to do, please it would prove that the `status` endpoint works.

> 2. doing a right click on Selenium Server 3.x would throw an exception

Yeah, I think it makes sense to test that too. Should be easy as well, probably you wouldn't even need to visit a page, even `about:blank` could work.

> Even if we add them, then how do I test on Selenium 4 so that only single test is executed?

In the driver test config, I would move `$majorSeleniumServerVersion = ...` to the top (it's a relatively fast operation, but you could also decide to cache it by setting it from constructor).

Then in `skipMessage()` something like:
```php
// change the class name and method name to whatever you do to test point 2 above
if ($this->majorSeleniumServerVersion >= 4 && [$testCase, $test] !== [CustomWebDriverTest::class, 'testThatSeleniumMustBeLessThan4']) {
    return 'Selenium must be less than version 4';
}
```
on the whole should be pretty standard...unless I missed something.